### PR TITLE
Fix val_progress_bar total with num_sanity_val_steps

### LINKED ISSUE #2891: The total number of batches shows by the progress bar of the sanity check is wrong
## 🐛 Bug
The `total` of the sanity check progress bar is set by
https://github.com/PyTorchLightning/pytorch-lightning/blob/4d0406ec8bf1c9147b34eb607411b78a9cd28243/pytorch_lightning/callbacks/progress.py#L296

The progress bar will always show `trainer.num_sanity_val_steps` even if  the length of the validation `DataLoader` is less than `trainer.num_sanity_val_steps`.

Maybe the `total` could be computed by
```python
from pytorch_lightning.trainer import data_loading

num_full_val_dataloader_batches = [
    len(dataloader) if data_loading._has_len(dataloader) else float('inf')
    for dataloader in trainer.val_dataloaders
]
self.val_progress_bar.total = convert_inf(
    sum(min(num_batches, trainer.num_sanity_val_steps)
            for num_batches in num_full_val_dataloader_batches))
```

We use the private function `data_loading._has_len` to check if `dataloader` has `__len__`, maybe we could make `data_loading._has_len` public. 

Or we could make `num_full_val_dataloader_batches` (and `num_full_train_dataloader_batches`) a member variable of `Trainer` and update the value in `pytorch_lightning.trainer.data_loading.TrainerDataLoadingMixin`. 

### To Reproduce

The progress bar of the sanity check in the following code (`num_sanity_val_steps == 999` and `len(val_data_loader) == 10`) shows
```
Validation sanity check:   1%|          | 9/999 [00:09<16:31,  1.00s/it]`
```

#### Code sample
```python
import time

import pytorch_lightning as pl
from torch.utils import data


class Dataset(data.Dataset):

  def __init__(self, length):
    self._elements = list(range(length))

  def __getitem__(self, item):
    return self._elements[item]

  def __len__(self):
    return len(self._elements)


class Model(pl.LightningModule):

  def forward(self, *args, **kwargs):
    pass

  def training_step(self, *args, **kwargs):
    pass

  def train_dataloader(self):
    pass

  def configure_optimizers(self):
    pass

  def validation_step(self, *args, **kwargs):
    time.sleep(1)
    return pl.EvalResult()


if __name__ == '__main__':
  model = Model()

  val_dataset_length = 10
  val_dataset = Dataset(val_dataset_length)
  val_data_loader = data.DataLoader(val_dataset)

  trainer = pl.Trainer(num_sanity_val_steps=999, limit_val_batches=999,
                       max_epochs=0)
  trainer.fit(model, val_dataloaders=val_data_loader)

```
### Expected behavior
The program above should be
```
Validation sanity check: 100%|██████████| 10/10 [00:10<00:00,  1.00s/it]
```
### Environment
* CUDA:
	- GPU:
	- available:
	- version:
* Packages:
	- numpy:             1.18.5
	- pyTorch_debug:     False
	- pyTorch_version:   1.6.0+cpu
	- pytorch-lightning: 0.9.0rc11
	- tensorboard:       1.15.0
	- tqdm:              4.48.2
* System:
	- OS:                Windows
	- architecture:
		- 64bit
		- WindowsPE
	- processor:
	- python:            3.7.3
	- version:           10.0.18362

### PR DESCRIPTION
**Pull request by rohitgr7**
## What does this PR do?

Fixes #2891 

# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [ ] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests? 
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Review by SkafteNicki — APPROVED** [review]

File: pytorch_lightning/callbacks/progress.py [hunk 1]
```diff
@@ -340,8 +343,9 @@ def on_train_batch_end(self, trainer, pl_module, batch, batch_idx, dataloader_id
 
     def on_validation_start(self, trainer, pl_module):
         super().on_validation_start(trainer, pl_module)
-        self.val_progress_bar = self.init_validation_tqdm()
-        self.val_progress_bar.total = convert_inf(self.total_val_batches)
+        if self._sanity_check_completed is True:
```
Comment (Borda) [inline pytorch_lightning/callbacks/progress.py:346]:
```suggestion
        if self._sanity_check_completed:
```

**Review by Borda — APPROVED** [review]
LGTM

File: tests/trainer/test_trainer.py [hunk 2]
```diff
@@ -949,7 +948,39 @@ def test_num_sanity_val_steps_neg_one(tmpdir, limit_val_batches):
...
+def test_num_sanity_val_steps_progress_bar(tmpdir, limit_val_batches, expected):
+    """
+    Test val_progress_bar total with "num_sanity_val_steps" Trainer argument.
+    """
+    class CustomCallback(Callback):
+        def __init__(self):
+            self.val_progress_bar_total = 0
+
+        def on_validation_epoch_end(self, trainer, pl_module):
+            self.val_progress_bar_total += trainer.progress_bar_callback.val_progress_bar.total
```
Comment (awaelchli) [inline tests/trainer/test_trainer.py:966]: I see a small issue with this test. 
This relies on the fact that the progress bar callback gets registered in the trainer in a "special" way. 
If we refine our callback system in the future, we will have to rewrite this test. I'll think of a better way, if there exists one :)

File: tests/trainer/test_trainer.py [hunk 3]
```diff
@@ -949,7 +948,39 @@ def test_num_sanity_val_steps_neg_one(tmpdir, limit_val_batches):
...
+
+    model = EvalModelTemplate()
+    cb = CustomCallback()
+
+    trainer = Trainer(
+        default_root_dir=tmpdir,
+        max_epochs=1,
+        num_sanity_val_steps=2,
+        limit_train_batches=0,
```
Comment (awaelchli) [inline tests/trainer/test_trainer.py:975]: I did not know we can set train batches to 0 :)

**Review by awaelchli — COMMENTED** [review]
Could you briefly explain why we need the new bool flag after refactors and not before? Thanks

File: pytorch_lightning/callbacks/progress.py (hunk 1 above)
Comment (rohitgr7) [reply pytorch_lightning/callbacks/progress.py:346]: it's the same thing.

**Comment by rohitgr7** [comment]
@awaelchli I don't remember what was the workflow before :sweat_smile: but since both sanity check and validation uses the same `val_progress_bar`, for sanity check it calls the `on_validation_start` hook in `run_evaluation` which re-registers the `val_progress_bar` with the number of validation batches. Partial workflow explained [here](https://github.com/PyTorchLightning/pytorch-lightning/issues/2891#issuecomment-682210162).

File: tests/trainer/test_trainer.py [hunk 4]
```diff
@@ -949,7 +948,39 @@ def test_num_sanity_val_steps_neg_one(tmpdir, limit_val_batches):
...
     )
     assert trainer.num_sanity_val_steps == float('inf')
-    val_dataloaders = model.val_dataloader__multiple()
+
+
+@pytest.mark.parametrize(['limit_val_batches', 'expected'], [
+    pytest.param(0, 0),
+    pytest.param(5, 7),
+])
+def test_num_sanity_val_steps_progress_bar(tmpdir, limit_val_batches, expected):
```
[the remark this answers is not in the document]
Comment (rohitgr7) [reply tests/trainer/test_trainer.py:957]: yeah, I think it's better to be there. I'll move it.

File: tests/trainer/test_trainer.py (hunk 2 above)
Comment (rohitgr7) [reply tests/trainer/test_trainer.py:966]: I can do this by overriding the `ProgressBar` instead of `Callback` if it sounds good.

Comment (awaelchli) [reply tests/trainer/test_trainer.py]: ok, sounds good!

File: pytorch_lightning/callbacks/progress.py (hunk 1 above)
Comment (williamFalcon) [inline pytorch_lightning/callbacks/progress.py:346]: these flags aren't needed...

we already have:

```python
if trainer.running_sanity_check:
```

**Review by williamFalcon — CHANGES_REQUESTED** [review]
Please don't add another variable...

use: if trainer.running_sanity_check:

File: pytorch_lightning/callbacks/progress.py (hunk 1 above)
Comment (rohitgr7) [reply pytorch_lightning/callbacks/progress.py]: didn't know about this. Updated!

**Review by awaelchli — APPROVED** [review]
nice

**Review by SkafteNicki — APPROVED** [review]
LGTM

**Comment by rohitgr7** [comment]
@williamFalcon is it good now??

**Comment by williamFalcon** [comment]
much better... but are those tests dropped now? for checking iterable loaders?

**Comment by rohitgr7** [comment]
Never added tests for iterable datasets here. Those are already present in the tests. This is just for tqdm progress bar total which was not working correctly for  sanity check after refs.

**Comment by williamFalcon** [comment]
image

These lines. wondering why the are deleted

**Comment by rohitgr7** [comment]
@williamFalcon these are redundant lines now since you removed some assertions here: https://github.com/PyTorchLightning/pytorch-lightning/pull/3123/commits/b9f4e7c8af1f6492f70bc60b9848eb6ce435f954, not sure why. But these were important tests for sanity checks. Should I add them back with the updated API call in this PR if required?

**Comment by williamFalcon** [comment]
ok got it. i see, thanks

**Comment by awaelchli** [comment]
> @williamFalcon these are redundant lines now since you removed some assertions here: b9f4e7c, not sure why. But these were important tests for sanity checks. Should I add them back with the updated API call in this PR if required?

yes please, if these were removed by accident, we need to add the assertions back. 
These checks make sure that `num sanity val steps` lead to the correct number of steps with multiple dataloaders.

**Comment by rohitgr7** [comment]
> yes please, if these were removed by accident, we need to add the assertions back.
These checks make sure that num sanity val steps lead to the correct number of steps with multiple dataloaders.

ok will add them back :v:

**Pull request merged**